crates/red_knot_python_semantic/src/types/class_base.rs

```diff
@@ -165,25 +167,36 @@ impl<'db> ClassBase<'db> {
                 KnownInstanceType::Callable => {
                     Self::try_from_type(db, todo_type!("Support for Callable as a base class"))
                 }
-                KnownInstanceType::Protocol => Some(ClassBase::Dynamic(DynamicType::TodoProtocol)),
+                KnownInstanceType::Protocol => Some(ClassBase::Protocol),
+                KnownInstanceType::Generic => Some(ClassBase::Generic),
             },
         }
     }
 
     pub(super) fn into_class(self) -> Option<ClassType<'db>> {
         match self {
             Self::Class(class) => Some(class),
-            Self::Dynamic(_) => None,
+            Self::Dynamic(_) | Self::Generic | Self::Protocol => None,
         }
     }
 
     /// Iterate over the MRO of this base
-    pub(super) fn mro(
-        self,
-        db: &'db dyn Db,
-    ) -> Either<impl Iterator<Item = ClassBase<'db>>, impl Iterator<Item = ClassBase<'db>>> {
+    pub(super) fn mro(self, db: &'db dyn Db) -> impl Iterator<Item = ClassBase<'db>> {
         match self {
-            ClassBase::Dynamic(_) => Either::Left([self, ClassBase::object(db)].into_iter()),
+            ClassBase::Protocol => Either::Left(Either::Left(
+                [self, ClassBase::Generic, ClassBase::object(db)].into_iter(),
+            )),
+            ClassBase::Dynamic(DynamicType::SubscriptedProtocol) => Either::Left(Either::Left(
+                [
+                    self,
+                    ClassBase::Dynamic(DynamicType::SubscriptedGeneric),
+                    ClassBase::object(db),
+                ]
+                .into_iter(),
+            )),
+            ClassBase::Dynamic(_) | ClassBase::Generic => {
+                Either::Left(Either::Right([self, ClassBase::object(db)].into_iter()))
+            }
```

Comment:
That's annoying that the different array lengths produce different types and therefore different `Either` wrappers. There's probably a way to simplify this down to a single type/wrapper using slices instead of arrays...but it's also probably not worth the effort! (So no recommended change here, just me blathering)

Reply:
I can't see a way of doing it with slices while keeping the borrow checker happy. But I think having to use nested `Either` types is reason enough to bite the bullet and write a custom `Iterator` type 😄 it's not _that_ much boilerplate, and it makes it easier to read at the callsite.

Reply:
FWIW I don't have strong feelings either way. Both versions seem kind of overwrought for what it's actually doing, but that's just life with the borrow checker sometimes :) The dedicated iterator is definitely more boilerplate overall, but also makes _this_ method less painful to read (so it encapsulates the pain a little better). I approved because I'm fine with either.

Reply:
Yeah, I agree that both versions seem annoyingly verbose/overwrought! I weakly prefer the version I have now, though, basically for the reasons Carl mentions. I also don't have strong feelings any which way, though.